sw/device/lib/crypto/impl/sha2/sha512.c

```diff
@@ -0,0 +1,336 @@
+  size_t partial_block_len =
+      (state->total_len.lower >> 3) % kSha512MessageBlockBytes;
+  memcpy(block.data, state->partial_block, partial_block_len);
+
+  // Initialize the context for the OTBN message buffer.
+  sha512_otbn_ctx_t ctx = {.num_blocks = 0};
+
+  // Process the message one block at a time, including partial data if it is
+  // present (which is only possible on the first iteration).
+  while (msg_len >= kSha512MessageBlockBytes - partial_block_len) {
+    size_t available_len = kSha512MessageBlockBytes - partial_block_len;
+    memcpy((unsigned char *)block.data + partial_block_len, msg, available_len);
+    msg += available_len;
+    msg_len -= available_len;
+    HARDENED_TRY(process_block(&ctx, &block));
+    partial_block_len = 0;
+  }
+
+  // Copy remaining mesage data into the working block (after partial data, if
+  // it is present). Because of the loop condition above, this must not be a
+  // full block.
+  memcpy((unsigned char *)block.data + partial_block_len, msg, msg_len);
+
+  // Add padding if necessary.
+  if (padding_needed == kHardenedBoolTrue) {
+    sha512_message_block_t additional_block;
+    size_t padding_len =
+        add_padding(new_state.total_len, &block, &additional_block);
+    // We always fill `block`; process it.
+    HARDENED_TRY(process_block(&ctx, &block));
+    // Check if `additional_block` was used, and process it if so.
+    if (padding_len >= kSha512MessageBlockBytes) {
+      HARDENED_TRY(process_block(&ctx, &additional_block));
+    }
+  }
+
+  // If there are any unprocessed blocks currently in DMEM, run the program one
+  // final time.
+  if (ctx.num_blocks > 0) {
+    HARDENED_TRY(process_message_buffer(&ctx));
+  }
+
+  // Read the final state from OTBN dmem. The state is still in the special
+  // form the OTBN app uses, with the 64-bit state words aligned to wide-word
+  // boundaries.
+  otbn_addr_t state_read_addr = kOtbnVarSha512State;
+  for (size_t i = 0; i + 1 < kSha512StateWords; i += 2) {
+    HARDENED_TRY(otbn_dmem_read(1, state_read_addr, &new_state.H[i + 1]));
+    HARDENED_TRY(
+        otbn_dmem_read(1, state_read_addr + sizeof(uint32_t), &new_state.H[i]));
+    state_read_addr += kOtbnWideWordNumBytes;
+  }
+
+  // Clear OTBN's memory.
+  HARDENED_TRY(otbn_dmem_sec_wipe());
+
+  // At this point, no more errors are possible; it is safe to update the
+  // context object.
+  memcpy(state, &new_state, sizeof(new_state));
+  memcpy(state->partial_block, block.data, sizeof(block.data));
+  return OTCRYPTO_OK;
+}
+
+status_t sha512_update(sha512_state_t *state, const uint8_t *msg,
+                       const size_t msg_len) {
+  // Process new data with no padding.
+  return process_message(state, msg, msg_len, kHardenedBoolFalse);
+}
+
+status_t sha512_final(sha512_state_t *state, uint8_t *digest) {
+  // Construct padding.
+  HARDENED_TRY(process_message(state, NULL, 0, kHardenedBoolTrue));
+  memcpy(digest, state->H, kSha512StateBytes);
```

Comment:
Are there any fields in `state` that need `hardened_memshred` and `memset -> 0`? Same for the one-shot function.

Reply:
Good point, I added a `state_shred` function that gets called after the digest is copied.